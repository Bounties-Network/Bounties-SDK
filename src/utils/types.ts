import BigNumber from "bignumber.js";


export interface HttpMethod {
    method: string,
    headers?: { [s: string]: string },
    withCredentials: boolean
}

export interface DjangoListResponse<T> {
    count: number,
    results: T[]
}

export interface PlatformQueryParams {
    platform?: string
}

export interface DjangoListQueryParams extends PlatformQueryParams {
    limit?: number
    offset?: number
}

export interface Metadata {
    platform: string
    schemaVersion: number
    schemaName: string
}

export enum Difficulty {
    beginner = 'Beginner',
    intermediate = 'Intermediate',
    advanced = 'Advanced'
}

export interface BountyData {
    title: string,
    body: string,
    categories: string[],
    expectedNumberOfRevisions: number,
    hasPrivateFulfillments: boolean,
    difficulty: Difficulty,
    attachments: {
        ipfsHash?: string,
        ipfsFileName?: string,
        url?: string,
    },
    metadata: Metadata
}

export interface MutableBountyData {
    data?: BountyData,
    approvers?: string[],
    controller?: string,
    deadline?: BigNumber
}