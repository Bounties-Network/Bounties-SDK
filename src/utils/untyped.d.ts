declare module 'ipfs-mini'
declare module 'truffle-hdwallet-provider'