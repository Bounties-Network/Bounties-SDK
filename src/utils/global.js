import config from '../config';

export let web3 = {};

let API_ENDPOINT;
//let API_ENDPOINT = 'http://localhost:8000';
// update this to be an env passthrough
export const apiEndpoint = {
  set: endpoint => (API_ENDPOINT = endpoint),
  get: () => API_ENDPOINT
};

export const expandPlatforms = platforms =>
  flatten(map(platform => config.platforms[platform], platforms)).join(',');