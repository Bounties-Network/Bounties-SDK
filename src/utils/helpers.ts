import { BigNumber } from 'bignumber.js';
import { Tx } from 'web3/eth/types';

export const calculateDecimals = (amount: BigNumber, decimals: number) => {
  const baseTen = new BigNumber(10, 10);
  const multiplyingDecimal = baseTen.exponentiatedBy(decimals);
  const amountBase = new BigNumber(amount, 10);
  return amountBase.times(multiplyingDecimal, 10).toString(10);
};

// export const readFile = file =>
//   new Promise((resolve, reject) => {
//     const reader = new window.FileReader();
//     reader.onloadend = () => {
//       resolve(reader);
//     };
//     reader.readAsArrayBuffer(file);
//   });

// export const promisify = inner =>
//   new Promise((resolve, reject) =>
//     inner((err, res) => {
//       if (err) {
//         reject(err);
//       } else {
//         resolve(res);
//       }
//     })
//   );

// export const proxiedWeb3Handler = {
//   // override getter
//   get: (target, name) => {
//     const inner = target[name];
//     if (inner instanceof Function) {
//       // Return a function with the callback already set.
//       return (...args) => promisify(cb => inner(...args, cb));
//     } else if (typeof inner === 'object') {
//       // wrap inner web3 stuff
//       return new Proxy(inner, proxiedWeb3Handler);
//     } else {
//       return inner;
//     }
//   }
// };

// export const promisifyContractCall = (contractFunction: Function, options: object) => (
//   ...args: any[]
// ) => {
//   return new Promise((resolve, reject) => {
//     return contractFunction(...args)
//       .send(options)
//       .on('transactionHash', hash => resolve(hash))
//       .on('error', error => reject(error));
//   });
// };

// export const batchContractMethods = (...methods) => {
//   return new Promise((resolve, reject) => {
//     const batch = new this._web3.eth.BatchRequest();
//     let completedCount = 0;
//     const txHashes = [];
//     each((method, idx) => {
//       const [fn, web3Options] = method;
//       batch.add(
//         fn.request(web3Options, (err, txHash) => {
//           if (err) {
//             return reject(err);
//           }
//           txHashes[idx] = txHash;
//           if (++completedCount === methods.length) {
//             return resolve(txHashes);
//           }
//         })
//       );
//     }, methods);
//     batch.execute();
//   });
// };

// export const addBuffer = (add: Function) => (filename: string, bufferContent: Buffer) =>
//   new Promise((resolve, reject) => {
//     // due to es5 issues - we load this via a CDN
//     // const ipfs = window.IpfsApi({
//     //   host: 'ipfs.infura.io',
//     //   port: 5001,
//     //   protocol: 'https'
//     // });

//     add(
//       [{ path: `/bounties/${filename}`, content: bufferContent }],
//       (err: Error, response: any) => {
//         if (err) {
//           reject(err);
//         }
//         resolve(response[1].hash);
//       }
//     )
//   })

export const addJSON = (addJSON: Function) => (data: { [s: string]: [any] }) =>
  new Promise((resolve, reject) => {
    addJSON(data, (err: Error, response: string) => {
      if (err) reject(err)
      resolve(response)
    })
  })

// export const handleOnChainAction = (action: (...args: any) => string) => {
//   return new Promise((resolve, reject) => {
//     try {
//       resolve(action())
//     } catch(e) {
//       reject(e)
//     }
//   })
// }