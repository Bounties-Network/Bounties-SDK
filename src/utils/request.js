import axios from 'axios';

import {
  HTTP_401_UNAUTHORIZED,
  HTTP_200_OK,
  HTTP_403_FORBIDDEN,
  HTTP_500_INTERNAL_SERVER_ERROR,
  HTTP_300_MULTIPLE_CHOICES
} from './constants';

const POST_OPTIONS = {
  method: 'POST',
  headers: {
    accept: 'application/json',
    'content-type': 'application/json'
  },
  withCredentials: true
};

const PUT_OPTIONS = {
  method: 'PUT',
  headers: {
    accept: 'application/json',
    'content-type': 'application/json'
  },
  withCredentials: true
};

const PATCH_OPTIONS = {
  method: 'PATCH',
  headers: {
    accept: 'application/json',
    'content-type': 'application/json'
  },
  withCredentials: true
};

const OPTIONS_OPTIONS = {
  method: 'OPTIONS',
  headers: {
    accept: 'application/json',
    'content-type': 'application/json'
  },
  withCredentials: true
};

const GET_OPTIONS = {
  method: 'GET',
  withCredentials: true
};

const OPTIONS = {
  'PUT': PUT_OPTIONS,
  'POST': POST_OPTIONS,
  'PATCH': PATCH_OPTIONS,
  'OPTIONS': OPTIONS_OPTIONS,
  'GET': GET_OPTIONS
}

function handleError(err) {
  const error = new Error();
  error.errorStatus = '';
  if (err.response) {
    const response = err.response;
    if (
      response.status === HTTP_401_UNAUTHORIZED ||
      response.status === HTTP_403_FORBIDDEN
    ) {
      // include redirect-type logic in here
    }

    if (response.status >= HTTP_500_INTERNAL_SERVER_ERROR) {
      error.errorStatus = response.status;
      error.errorMessage = response.statusText;
      throw error;
    }

    error.errorStatus = response.status;
  }

  error.errorMessage = err.message;
  throw error;
}

function checkRequestStatus(response) {
  if (
    response.status >= HTTP_200_OK &&
    response.status < HTTP_300_MULTIPLE_CHOICES
  ) {
    return response.data;
  }
}

export default class Request {
  constructor(bounties) {
    this._bounties = bounties
    this._endpoint = bounties._endpoint
  }

  request(url, endpoint, method='GET', options, customErrorHandler) {
    const requestUrl = url.slice(0, 4) === 'http' ? url : `${endpoint}/${url}`;
    return axios
      .request(requestUrl, { ...OPTIONS[method.toUpperCase()], ...options })
      .then(checkRequestStatus)
      .catch(handleError);
  }
 
  get(url, query, customErrorHandler) { 
    return this.request(url, this._endpoint, 'GET', { params: query }, customErrorHandler) 
  }
  
  post(url, data, customErrorHandler) { 
    return this.request(url, this._endpoint, 'POST', { params: data }, customErrorHandler) 
  }
}
