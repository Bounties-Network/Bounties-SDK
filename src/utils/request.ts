import axios, { AxiosRequestConfig, AxiosPromise, AxiosResponse } from 'axios';
import Bounties from '../bounties';
import {
    HTTP_401_UNAUTHORIZED,
    HTTP_200_OK,
    HTTP_403_FORBIDDEN,
    HTTP_500_INTERNAL_SERVER_ERROR,
    HTTP_300_MULTIPLE_CHOICES
} from './constants';


export class Request {
    _endpoint: string

    constructor(endpoint: string) {
        this._endpoint = endpoint
    }

    request(endpoint: string, relativePath: string, method: string = 'GET', options: AxiosRequestConfig): AxiosPromise {
        return axios
            .request({ url: `${endpoint}/${relativePath}`, ...HttpOptions[method.toUpperCase()], ...options })
            .then(checkRequestStatus)
            .catch(handleError);
    }

    get(url: string, query?: object): AxiosPromise {
        return this.request(this._endpoint, url, 'GET', { params: query })
    }

    post(url: string, data: object): AxiosPromise {
        return this.request(this._endpoint, url, 'POST', { params: data })
    }
}

export const HttpOptions: { [s: string]: HttpMethod } = {
    'PUT': {
        method: 'PUT',
        headers: {
            accept: 'application/json',
            'content-type': 'application/json'
        },
        withCredentials: true
    },
    'POST': {
        method: 'POST',
        headers: {
            accept: 'application/json',
            'content-type': 'application/json'
        },
        withCredentials: true
    },
    'PATCH': {
        method: 'PATCH',
        headers: {
            accept: 'application/json',
            'content-type': 'application/json'
        },
        withCredentials: true
    },
    'OPTIONS': {
        method: 'OPTIONS',
        headers: {
            accept: 'application/json',
            'content-type': 'application/json'
        },
        withCredentials: true
    },
    'GET': {
        method: 'GET',
        withCredentials: true
    } 
}

// The type binds provided by Axios do not include 
// message or statusCode so they are inlined here.
interface AxiosError {
    config: AxiosRequestConfig,
    code?: string,
    request?: any,
    response?: AxiosResponse,
    statusCode?: number,
    message?: string
}

function handleError(err: AxiosError) {
    if (err.statusCode) {
        throw new Error(`${err.statusCode}`)
    }

    throw err;
}

function checkRequestStatus(response: AxiosResponse) {
    if (
        response.status >= HTTP_200_OK &&
        response.status < HTTP_300_MULTIPLE_CHOICES
    ) {
        return response.data;
    }
}
