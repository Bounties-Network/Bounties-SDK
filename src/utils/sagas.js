export function* getTokenData(tokenAddress) {
  let symbol, decimals;
  try {
    const { tokenContract: tokenContractClient } = yield call(
      getTokenClient,
      tokenAddress
    );
    symbol = yield call(tokenContractClient.symbol().call);
    decimals = yield call(tokenContractClient.decimals().call);
  } catch (e) {
    const { tokenContract: tokenContractClient } = yield call(
      getTokenClient,
      tokenAddress,
      'DSToken'
    );
    symbol = yield call(tokenContractClient.symbol().call);
    decimals = yield call(tokenContractClient.decimals().call);
  }
  return {
    symbol,
    decimals
  };
}