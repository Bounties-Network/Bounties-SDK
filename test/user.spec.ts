import Web3 from 'web3'
import HDWalletProvider from 'truffle-hdwallet-provider'
import BigNumber from 'bignumber.js';
import { map } from 'lodash';

import { bounties, bountyContract } from './utils/instances'
import { FakeHttpProvider } from './utils/fakeHttpProvider'
import { constants } from './utils/constants'
import { buildAddress, encodeParameters, sanitize } from './utils/helpers'


const {
    USER_ADDRESS,
    FACTORY_ADDRESS,
    BOUNTY_ADDRESS,
    GAS,
    GAS_PRICE,
    TX_HASH,
    IPFS_HASH,
    TX_OPTIONS,
    BOUNTY_DATA,
    USER_FIELDS_TO_SANITIZE,
    STAT_FIELDS_TO_SANITIZE
} = constants

describe('user resource', () => {
    beforeAll(() => {
        const mnemonic = 'candy maple cake sugar pudding cream honey rich smooth crumble sweet treat'
        const provider = new HDWalletProvider(mnemonic, "https://rinkeby.infura.io")
        bounties._web3.setProvider(provider)
    })

    describe('offchain', () => {
        describe('retrieve', () => {
            it('should retrieve user', async () => {
                const user = bounties.user.retrieve(USER_ADDRESS)
                await expect(user).resolves.toEqual(expect.anything())
                const sanitized = sanitize(
                    await user,
                    [...USER_FIELDS_TO_SANITIZE, ...STAT_FIELDS_TO_SANITIZE]
                )
                expect(sanitized).toMatchSnapshot()
            })

            it('should not return user when the address does not exist', async () => {
                const user = await bounties.user.retrieve('1234')
                // TODO: promise should reject if no user is found
                // https://github.com/Bounties-Network/BountiesAPI/issues/252
                expect(user).toMatchSnapshot()
            })
        })

        describe('nonce', () => {
            it('should fetch user\'s nonce', async () => {
                const nonce = bounties.user._nonce(USER_ADDRESS)
                await expect(nonce).resolves.toEqual(expect.anything())
                const sanitized = sanitize(await nonce, ['nonce'])
                expect(sanitized).toMatchSnapshot()
            })
        })

        describe('login', () => {
            it('should log a user in', async () => {
                const user = bounties.user.login()
                await expect(user).resolves.toEqual(expect.anything())
                const sanitized = sanitize(
                    await user,
                    [...USER_FIELDS_TO_SANITIZE, 'hasSignedUp']
                )
                expect(sanitized).toMatchSnapshot()
            })
        })

        describe('logout', () => {
            it('should logout a user', async () => {
                const response = bounties.user.logout()
                await expect(response).resolves.toMatchSnapshot()
            })
        })
    })
})