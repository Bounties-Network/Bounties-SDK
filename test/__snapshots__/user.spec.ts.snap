// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user resource offchain login should log a user in 1`] = `
Object {
  "hasSignedUp": "[SANITIZED]",
  "user": Object {
    "categories": "[SANITIZED]",
    "created": "2018-12-11T12:37:20.336958",
    "dismissed_signup_prompt": null,
    "dribble": "[SANITIZED]",
    "edited": "[SANITIZED]",
    "email": "[SANITIZED]",
    "ens_domain": "[SANITIZED]",
    "github": "[SANITIZED]",
    "id": 2304,
    "languages": "[SANITIZED]",
    "large_profile_image_url": "[SANITIZED]",
    "last_logged_in": "[SANITIZED]",
    "last_viewed": "[SANITIZED]",
    "linkedin": "[SANITIZED]",
    "name": "[SANITIZED]",
    "organization": "[SANITIZED]",
    "page_preview": "[SANITIZED]",
    "public_address": "0x627306090abab3a6e1400e9345bc60c78a8bef57",
    "settings": "[SANITIZED]",
    "skills": "[SANITIZED]",
    "small_profile_image_url": "[SANITIZED]",
    "twitter": "[SANITIZED]",
    "wants_marketing_emails": "[SANITIZED]",
    "website": "[SANITIZED]",
  },
}
`;

exports[`user resource offchain logout should logout a user 1`] = `"Success"`;

exports[`user resource offchain nonce should fetch user's nonce 1`] = `
Object {
  "has_signed_up": true,
  "nonce": "[SANITIZED]",
}
`;

exports[`user resource offchain retrieve should not return user when the address does not exist 1`] = `
Object {
  "stats": Object {},
  "user": null,
}
`;

exports[`user resource offchain retrieve should retrieve user 1`] = `
Object {
  "stats": Object {
    "awarded": "[SANITIZED]",
    "earned": "[SANITIZED]",
    "fulfiller_fulfillment_acceptance": "[SANITIZED]",
    "fulfiller_ratings_given": null,
    "fulfiller_ratings_received": "[SANITIZED]",
    "issuer_fulfillment_acceptance": "[SANITIZED]",
    "issuer_ratings_given": "[SANITIZED]",
    "issuer_ratings_received": "[SANITIZED]",
    "total_bounties": "[SANITIZED]",
    "total_fulfillments": "[SANITIZED]",
    "total_fulfillments_on_bounties": "[SANITIZED]",
  },
  "user": Object {
    "categories": "[SANITIZED]",
    "created": "2018-10-19T00:07:13.243450",
    "dismissed_signup_prompt": null,
    "dribble": "[SANITIZED]",
    "edited": "[SANITIZED]",
    "email": "[SANITIZED]",
    "ens_domain": "[SANITIZED]",
    "github": "[SANITIZED]",
    "id": 845,
    "languages": "[SANITIZED]",
    "large_profile_image_url": "[SANITIZED]",
    "last_logged_in": "[SANITIZED]",
    "last_viewed": "[SANITIZED]",
    "linkedin": "[SANITIZED]",
    "name": "[SANITIZED]",
    "organization": "[SANITIZED]",
    "page_preview": "[SANITIZED]",
    "public_address": "0xe23f91725c6c18204743592f963688b8b8dc2ced",
    "settings": "[SANITIZED]",
    "skills": "[SANITIZED]",
    "small_profile_image_url": "[SANITIZED]",
    "twitter": "[SANITIZED]",
    "wants_marketing_emails": "[SANITIZED]",
    "website": "[SANITIZED]",
  },
}
`;
